python/ql/src/semmle/python/regex.qll

```diff
@@ -39,24 +44,78 @@ private Value relevant_re_attr(string name) {
 predicate used_as_regex(Expr s, string mode) {
   (s instanceof Bytes or s instanceof Unicode) and
   /* Call to re.xxx(regex, ... [mode]) */
-  exists(CallNode call, string name |
-    call.getArg(0).pointsTo(_, _, s.getAFlowNode()) and
-    call.getFunction().pointsTo(relevant_re_attr(name))
+  exists(DataFlow::CallCfgNode call, string name |
+    call.getArg(0).asExpr() = s and
+    call = relevant_re_member(name).getACall()
   |
     mode = "None"
     or
-    exists(Value obj | mode = mode_from_mode_object(obj) |
-      exists(int flags_arg |
-        re_module_function(name, flags_arg) and
-        call.getArg(flags_arg).pointsTo(obj)
-      )
-      or
-      call.getArgByName("flags").pointsTo(obj)
+    mode = mode_from_node([call.getArg(re_member_flags_arg(name)), call.getArgByName("flags")])
+  )
+}
+
+/**
+ * Gets the canonical name for the API graph node corresponding to the `re` flag `flag`. For flags
+ * that have multiple names, we pick the long-form name as a canonical representative.
+ */
+private string canonical_name(API::Node flag) {
+  result in ["ASCII", "IGNORECASE", "LOCALE", "UNICODE", "MULTILINE", "TEMPLATE"] and
+  flag = API::moduleImport("re").getMember([result, result.prefix(1)])
+  or
+  flag = API::moduleImport("re").getMember(["DOTALL", "S"]) and result = "DOTALL"
+  or
+  flag = API::moduleImport("re").getMember(["VERBOSE", "X"]) and result = "VERBOSE"
+}
+
+/**
+ * A type tracker for regular expression flag names. Holds if the result is a node that may refer
+ * to the `re` flag with the canonical name `flag_name`
+ */
+private DataFlow::LocalSourceNode re_flag_tracker(string flag_name, DataFlow::TypeTracker t) {
+  t.start() and
+  exists(API::Node flag | flag_name = canonical_name(flag) and result = flag.getAUse())
+  or
+  exists(BinaryExprNode binop, DataFlow::Node operand |
+    operand.getALocalSource() = re_flag_tracker(flag_name, t.continue()) and
```

Comment:
I have a bit of trouble understanding if `t.continue()` is the only right choice here. 

I would have thought that if we were able to track a flag with some type tracker `t2`, if that flag is used in a binary or operation, the resulting type-tracker would be the continuation of `t2`. So

```codeql
  exists(BinaryExprNode binop, DataFlow::Node operand, DataFlow::TypeTracker t2 |
    operand.getALocalSource() = re_flag_tracker(flag_name, t2) and
    t = t2.continue() and
    ...
```

But I'm also wondering if we _need_ this continue stuff at all, or we could just use `re_flag_tracker(flag_name)` instead -- and if there would be any bad consequences of that. 

```codeql
  exists(BinaryExprNode binop, DataFlow::Node operand |
    operand.getALocalSource() = re_flag_tracker(flag_name) and
    ...
```

Reply:
Your suggested rewrite would have at least two consequences, both of which may affect performance, and one of which will affect behaviour:
1. By referring to `re_flag_tracker/1`, the fixpoint computation now has to evaluate both this and `re_flag_tracker/2` at the same time. Currently, `re_flag_tracker/1` is simply an extra join on top of the result of the other predicate. The impact on performance probably isn't terribly big, but I imagine there is _some_ overhead in doing this.
2. Probably more impactful on performance is the fact that by _not_ reusing the type tracker, we lose track of whether we have previously propagated the type information across a call. Thus, with your suggestion we might track into a call, then through a binary operation and then _out_ of a different call to the same function. This is potentially a much larger set of nodes.

Finally, rewriting to use `t2` actually doesn't change the behaviour. `t = t2.continue()` is equivalent to `t = t2 and t.attr = ""` (with a slight abuse of notation). Thus, `continue` really just checks that we're not tracking an attribute (which makes sense -- we can add `re` flags, but not objects that happen to have an `re` flag in an attribute).

Reply:
:+1:

> Finally, rewriting to use t2 actually doesn't change the behaviour. t = t2.continue() is equivalent to t = t2 and t.attr = "" (with a slight abuse of notation). Thus, continue really just checks that we're not tracking an attribute (which makes sense -- we can add re flags, but not objects that happen to have an re flag in an attribute).

Right. Although the behavior ends up being the same, for me, it just reads wrong semantically. I would like to use the `t2` approach, and have made a suggestion for that (couldn't do it in this thread, since that only covered one of the lines :facepalm:)

Reply:
Everywhere else (in Python and JavaScript) where this construction appears, and I do mean _everywhere_ is in the form `t.continue()`. We only see `t2.continue()` (and currently only in JavaScript) when `(back)track(t2,t)` is used in an adjacent disjunct.

So I think to suddenly change what is the standard idiom doesn't make much sense.

Rather, it seems to me that a better solution would be to add `predicate canContinue` to `TypeTracker`, with the meaning that `this.attr` is empty. Then we can reuse `t` in both places, as long as we make sure to check `... and t.canContinue()` and we don't have to waste precious space on declaring a variable that is exactly equal to `t` anyway.

This, however, I think it outside the scope of the present PR.

Reply:
I agree that changing the way it's used probably isn't in the scope of this PR.